### PR TITLE
Add option to not dump first and last timestep

### PR DESCRIPTION
**Pull request by NeilZaim**
For full diagnostics (including checkpoints), we currently always dump the first and last timestep regardless of the content of the `IntervalsParser`. This can be annoying in some situations (see e.g. #1495) so I propose to change this behaviour as follows:

- For the first timestep, dump only if asked by the `IntervalsParser`, which is the case most of the time (i.e. `diag.intervals = 10` will dump at timestep 0 but `diag.intervals = 10::10` will not dump anymore).

- The final timestep is still dumped by default regardless of the `IntervalsParser` but I've added a flag ``<diag_name>.dump_last_timestep`` that can be set to 0 to disable that.

Of course this is completely open to discussion and I really don't have a strong opinion on that.

Regarding the implementation: currently the diagnostic function `FilterComputePackFlush` has an optional boolean argument `force_flush` defaulting the `false`. This function is only called with `force_flush = true` at the first and last timestep.
I have thus removed this optional argument when calling the function in the first timestep in `WarpXInitData.cpp`.

So now `force_flush` is only used at the last timestep and I've thus renamed it `last_timestep` to make things more explicit. But maybe there were other intents for `force_flush` in which case we can find another implementation. What do you think @MaxThevenet ?

### CODE REVIEW

**Comment by MaxThevenet** [comment]
Thank you for this PR @NeilZaim. For the full diagnostics, I agree with what you say. In fact, could we apply what you propose to the last time step too?

As for the naming, I agree that, for FullDiagnostics, `force_flush` will only be applied at the last time step, so both names `force_flush` and `last_timestep` are equivalent. Back-transformed diagnostics (`BTDiagnostics`) work a bit differently: due to the Lorentz transform (z,t) -> (z',t') in these diagnostics, all output files (called `snapshots`) are initialized at the beginning of the simulation (one per output time in the lab frame, while the simulation runs in the boosted frame). Each time step in the simulation contains data that has to be written in each snapshot, approximately. To avoid uninterrupted data transfers, this data is gathered into buffers (one buffer per snapshot), and written to file only when the buffer is full. The problem is that, at the end of the simulation, if a buffer is not full (and they have no reason to be) it will not be dumped, and the data will be lost. For this reason, at the last time step, we force-flush all buffers to the snapshots. Hence, for BTD, I am not sure `last_timestep` is more accurate, because it is indeed used at the last simulation timestep, but it effectively fills a bit of all snapshots. Besides, I think it would make sense to have the option to force-flush periodically these buffers because, if we don't, we have to wait for the whole simulation to complete successfully before even the first snapshot to be complete.

Long story short: `force_flush` describes the function behaviour while `last_timestep` describes the function usage. As `Diagnostics` is a virtual class whose derived classes can behave quite differently, I think it is safer to not presuppose the usage. Hence, I'd recommend keeping `force_flush` as a name.

**Comment by NeilZaim** [comment]
@MaxThevenet Thanks for your reply! I completely understand your point regarding the variable naming. My main issue with `force_flush` was that, with the current status of this PR, we do not actually force flushing the FullDiagnostic at the last timestep (we only flush if `diag.dump_last_timestep = 1`). But this could change depending on what we decide regarding your first question:

> In fact, could we apply what you propose to the last time step too?

So I guess we can first decide what to do on this point and then find a suitable implementation. I think that for the last timestep we have 3 options.

- 1) Dump final timestep only if it is included in the intervals of the input file.
- 2) Same as i. but with a flag that can be used to force dumping the final timestep even if it isn't in the intervals
- 3) Same as ii. but the flag is true by default.

I initially chose option iii. in order to mimic the current behaviour of the code (in most cases we have `diag.intervals = integer_period`, in which case we dump timestep 0 and with solution iii. we would also dump the final timestep, regardless of whether it can be divided by the diag period). But mimicking the current behaviour is not a very strong reason to chose iii. so I'm also fine with i. and ii.. From what you said, you'd prefer to choose option i., is that correct?

**Comment by MaxThevenet** [comment]
You're right, one difference between first and last time step is that we know the first time step (0), but we don't always know the last one (the user can pass a physical time, for instance). However, it would be convenient if dumping the last time step could also be handled by the `IntervalsParser`, just to have all in one place. For instance, we could introduce a convention that `-1` in the interval parser is the last time step. Then the intervals parser could read `:-1:10` or `10:10:100,-1`. In that sense, first and last time steps are both handled by the `IntervalsParser`. What do you think?

**Comment by NeilZaim** [comment]
@MaxThevenet Yes I like the idea of using negative numbers so that -1 means last timestep, -2 means second to last timestep, etc. And even if we decide to keep a flag to dump the last timestep, I guess it's better to read this flag as we initialize the IntervalsParser, so that everything is handled there.

The main issue I see with using -1 is that currently a negative period means that we disable the feature. This will change, and there will be no error if someone thinks he is disabling the feature but in fact activates it only at the final timestep. But overall I think that since we're currently making non backward compatible changes as we unify the intervals notation (#1485), it might be a good time to do this change as well.

**Comment by MaxThevenet** [comment]
You're right, that is a significant change... @ax3l @dpgrote @RemiLehe should we do it anyway (you don't need to read the full conversation, just [this message](https://github.com/ECP-WarpX/WarpX/pull/1530#issuecomment-739937609) and [this one](https://github.com/ECP-WarpX/WarpX/pull/1530#issuecomment-741949190))? We could:
- Introduce `-1` for the last time step, `-2` for the previous one etc. That's very Python-like but breaks backward compatibility. We also need to still provide a way for the user to dump no output, so we could introduce a new convention, for instance use `plot_int = none` to have no output (the previous `-1`).
- Introduce `last` for the last timestep, so the user would write `:last:10` or `10:10:100,last` if we want to keep backward-compatibility.

**Comment by NeilZaim** [comment]
> We also need to still provide a way for the user to dump no output, so we could introduce a new convention, for instance use plot_int = none to have no output

For this specific point, using a 0 period would still work (`diag.intervals = 0` would still disable the diag)

**Comment by NeilZaim** [comment]
After discussions, we decided to:
- Dump first timestep only if included in the `IntervalsParser`
- Dump last timestep by default but with a flag to turn this off
- Implement negative numbers in the parser in a follow-up PR

In the end this is similar to the current status of the PR. I still think the PR requires some changes because I would prefer the flag `diag.dump_last_timestep` to be read as we initialize the `IntervalsParser`, which is currently not the case.

**Comment by NeilZaim** [comment]
Hi @MaxThevenet @RemiLehe . I'm reviving this old (and forgotten) PR. With the latest commit, it should implement what we had decided to do in a February dev meeting:

- Dump first timestep only if included in the IntervalsParser
- Dump last timestep by default but with a flag (labeled here `<diag_name>.dump_last_timestep`) to turn this off

For the implementation, I've added a function `MultiDiagnostics::FilterComputePackFlushLastTimestep` that loops over each diag and checks the value of the flag `dump_last_timestep` If true, this calls `FilterComputePackFlush` with `force_flush` set to `true`. Does that look good to you?

**Review by RemiLehe — APPROVED** [review]

**Comment by RemiLehe** [comment]
This looks good to me, thanks!
